include/circt/Dialect/HW/HWAttributes.td

```diff
@@ -299,6 +299,36 @@ def EnumFieldAttr : AttrDef<HWDialect, "EnumField"> {
   }];
 }
 
+// An attribute to define the wrapping interface for an external module.
+def ESIWrapAttr : AttrDef<HWDialect, "ESIWrap"> {
```

Comment:
Do we need a custom attribute for this? Is a dictionary attribute not suitable? I've usually seen custom attributes associated with dialects - however, i suspect that this is going to be a custom attribute/state associated with a set of passes, hence it shouldn't be a dialect attribute.

Reply:
I changed it to a DictionaryAttribute with the name esiWrap. It now looks like this:
'''
module {
  hw.module.extern @handshake_external_instance_ext_add(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>) attributes {esiWrap = {target = @ext_add, type = "combinational"}}
  hw.module @ext_add(%a: i32, %b: i32) -> (out: i32) {
    %0 = comb.add %a, %b : i32
    hw.output %0 : i32
  }
  hw.module @main(%a: !esi.channel<i32>, %b: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i0>) {
    %handshake_external_instance0.out0 = hw.instance "handshake_external_instance0" @handshake_external_instance_ext_add(in0: %a: !esi.channel<i32>, in1: %b: !esi.channel<i32>, clock: %clock: i1, reset: %reset: i1) -> (out0: !esi.channel<i32>)
    hw.output %handshake_external_instance0.out0, %ctrl : !esi.channel<i32>, !esi.channel<i0>
  }
}
'''

Reply:
And I completely removed the attribute type hw::ESIWrap since it not used

Reply:
```mlir
module {
  hw.module.extern @handshake_external_instance_ext_add(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>) attributes {esiWrap = {target = @ext_add, type = "combinational"}}
  hw.module @ext_add(%a: i32, %b: i32) -> (out: i32) {
    %0 = comb.add %a, %b : i32
    hw.output %0 : i32
  }
  hw.module @main(%a: !esi.channel<i32>, %b: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i0>) {
    %handshake_external_instance0.out0 = hw.instance "handshake_external_instance0" @handshake_external_instance_ext_add(in0: %a: !esi.channel<i32>, in1: %b: !esi.channel<i32>, clock: %clock: i1, reset: %reset: i1) -> (out0: !esi.channel<i32>)
    hw.output %handshake_external_instance0.out0, %ctrl : !esi.channel<i32>, !esi.channel<i0>
  }
}
```